packages/viz/src/lib/interactivity/viewport-features/ViewportFeaturesGenerator.ts

```diff
@@ -0,0 +1,167 @@
+import { Deck, Viewport } from '@deck.gl/core';
+import { MVTLayer } from '@deck.gl/geo-layers';
+import { Matrix4 } from '@math.gl/core';
+import { selectPropertiesFrom } from '../../utils/object';
+import { AggregationTypes, applyAggregations } from './aggregations';
+import { ViewportTile } from '../../../declarations/deckgl';
+import { GeometryData, ViewportFrustumPlanes } from './geometry/types';
+import { checkIfGeometryIsInsideFrustum } from './geometry/check';
+import {
+  getTransformationMatrixFromTile,
+  transformGeometryCoordinatesToCommonSpace
+} from './geometry/transform';
+
+const DEFAULT_OPTIONS = {
+  uniqueIdProperty: 'cartodb_id'
+};
+
+const DEFAULT_GET_FEATURES_OPTIONS = {
+  properties: [],
+  aggregations: {}
+};
+
+export class ViewportFeaturesGenerator {
+  private deckInstance: Deck | undefined;
+  private deckLayer: MVTLayer<string> | undefined;
+  private uniqueIdProperty: string;
+
+  constructor(
+    deckInstance?: Deck,
+    deckLayer?: MVTLayer<string>,
+    options: ViewportFeaturesGeneratorOptions = DEFAULT_OPTIONS
+  ) {
+    const { uniqueIdProperty = DEFAULT_OPTIONS.uniqueIdProperty } = options;
+
+    this.deckInstance = deckInstance;
+    this.deckLayer = deckLayer;
+    this.uniqueIdProperty = uniqueIdProperty;
+  }
+
+  isReady() {
+    return Boolean(this.deckInstance) && Boolean(this.deckLayer);
+  }
+
+  getFeatures(options: ViewportFeaturesOptions = DEFAULT_GET_FEATURES_OPTIONS) {
+    const {
+      properties = DEFAULT_GET_FEATURES_OPTIONS.properties,
+      aggregations = DEFAULT_GET_FEATURES_OPTIONS.aggregations
+    } = options;
+
+    const selectedTiles = this.getSelectedTiles();
+
+    const features = this.getViewportFilteredFeatures(
+      selectedTiles,
+      properties
+    );
+
+    return {
+      features,
+      aggregations: applyAggregations(features, aggregations)
+    };
+  }
+
+  private getViewportFilteredFeatures(
+    selectedTiles: ViewportTile[],
+    properties: string[]
+  ) {
+    const currentFrustumPlanes = this.getFrustumPlanes();
+    const featureCache = new Set<number>();
+
+    return selectedTiles
+      .map(tile => {
+        const transformationMatrix = getTransformationMatrixFromTile(tile);
+        const features = tile.content || [];
+
+        const featuresWithinViewport = features.filter(feature => {
```

Comment:
Nice function. It made me think about filtering the features client-side on a condition (like feature.pop_max > 50), and the widgets interaction

Reply:
Yes, that's possible with this!